ckanext/datagovuk/upload.py

```diff
@@ -0,0 +1,150 @@
+'''
+upload.py
+
+Contains functions that upload the organograms to S3.
+
+'''
+import cgi
+import os
+import StringIO
+import zipfile
+import mimetypes
+import collections
+import logging
+import datetime
+from dateutil import parser
+
+from slugify import slugify
+from pylons import config
+import boto3
+import requests
+from botocore.exceptions import ClientError
+
+import paste.fileapp
+import ckan.plugins.toolkit as toolkit
+import ckan.lib.uploader as uploader
+from ckan.common import request
+
+
+def setup_s3_bucket():
+    '''
+    setup_s3_bucket - Grabs the required info from config file and initializes S3 connection
+    '''
+    aws_access_key_id = config.get('ckan.datagovuk.s3_aws_access_key_id')
+    aws_secret_access_key = config.get('ckan.datagovuk.s3_aws_secret_access_key')
+    aws_region_name = config.get('ckan.datagovuk.s3_aws_region_name')
+    if aws_region_name:
+        s3 = boto3.resource('s3',
+                            aws_access_key_id=aws_access_key_id,
+                            aws_secret_access_key=aws_secret_access_key,
+                            region_name=aws_region_name)
+    else:
+        s3 = boto3.resource('s3',
+                            aws_access_key_id=aws_access_key_id,
+                            aws_secret_access_key=aws_secret_access_key)
+
+    bucket_name = config.get('ckan.datagovuk.s3_bucket_name')
+    bucket = s3.Bucket(bucket_name)
+
+    return bucket
+
+
+def upload_resource_to_s3(context, resource):
+    '''
+    upload_resource_to_s3
+
+    Uploads resource to S3 and modifies the following resource fields:
+    - 'upload'
+    - 'url_type'
+    - 'url'
+    '''
+
+    # Init logger
+    logger = logging.getLogger(__name__)
+    logger.info("Starting upload_resource_to_s3 for resource %s" % resource.get('name', ''))
+
+    # Init connection to S3
+    bucket = setup_s3_bucket()
+
+    # Get content type and extension
+    content_type, _ = mimetypes.guess_type(
+        resource.get('url', ''))
+    extension = mimetypes.guess_extension(content_type)
+
+    # Upload to S3
+    timestamp = datetime.datetime.utcnow()
+    pkg = toolkit.get_action('package_show')(context, {'id': resource['package_id']})
+    s3_filepath = (pkg.get('name')
+                   + '/'
+                   + 'resources'
+                   + '/'
+                   + resource.get('timestamp', timestamp.strftime("%Y-%m-%dT%H-%M-%SZ"))
+                   + '-'
+                   + slugify(resource.get('name'), to_lower=True)
+                   + extension)
```

Comment:
I wonder if this reads a bit better:

```python
filename = resource.get('timestamp', timestamp.strftime("%Y-%m-%dT%H-%M-%SZ")) \
    + '-' + slugify(resource.get('name'), to_lower=True) \
    + extension

s3_filepath = "/".join([pkg.get('name'), 'resources', filename])
```

Probably more personal preference at this point, but I thought it was worth mentioning.